MapboxNavigation/Cache.swift

```diff
@@ -0,0 +1,139 @@
+import Foundation
+
+public typealias CompletionHandler = () -> Void
+
+/**
+ A cache consists of both in-memory and on-disk components, both of which can be reset.
+ */
+@objc(MBBimodalCache)
+public protocol BimodalCache {
+    func clearMemory()
+    func clearDisk(completion: CompletionHandler?)
+}
+
+/**
+ A cache which supports storing images
+ */
+@objc(MBBimodalImageCache)
+public protocol BimodalImageCache: BimodalCache {
+    func store(_ image: UIImage, forKey key: String, toDisk: Bool, completion completionBlock: CompletionHandler?)
+    func image(forKey: String?) -> UIImage?
+}
+
+/**
+ A cache which supports storing data
+ */
+@objc(MBBimodalDataCache)
+public protocol BimodalDataCache: BimodalCache {
+    func store(_ data: Data, forKey key: String, toDisk: Bool, completion completionBlock: CompletionHandler?)
+    func data(forKey: String?) -> Data?
+}
+
+ /**
+  A general purpose on-disk cache used by both the ImageCache and DataCache implementations
+ */
+internal class FileCache {
+    let diskCacheURL: URL = {
+        let fileManager = FileManager.default
+        let basePath = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first!
+        let identifier = Bundle.mapboxNavigation.bundleIdentifier!
+        return basePath.appendingPathComponent(identifier + ".downloadedFiles")
+    }()
+
+    let diskAccessQueue = DispatchQueue(label: Bundle.mapboxNavigation.bundleIdentifier! + ".diskAccess")
+    var fileManager: FileManager?
+
+    init() {
+        diskAccessQueue.sync {
+            fileManager = FileManager()
+        }
+    }
+
+    public func store(_ data: Data, forKey key: String, completion: CompletionHandler?) {
+        let dispatchCompletion = {
+            if let completion = completion {
+                DispatchQueue.main.async {
+                    completion()
+                }
+            }
+        }
+        guard let fileManager = fileManager else {
+            dispatchCompletion()
+            return
+        }
+        let cacheURL = diskCacheURL
+        diskAccessQueue.async {
+            self.createCacheDirIfNeeded(cacheURL, fileManager: fileManager)
+            let cacheURL = self.cacheURLWithKey(key)
+
+            do {
+                try data.write(to: cacheURL)
+            } catch {
+                NSLog("================> Failed to write data to URL \(cacheURL)")
+            }
+            dispatchCompletion()
+        }
+
+    }
+
+    public func dataFromFileCache(forKey key: String?) -> Data? {
+        guard let key = key else {
+            return nil
+        }
+        do {
+            return try Data.init(contentsOf: cacheURLWithKey(key))
```

Comment:
Is there an async way to read from file cache using `diskAccessQueue` or similar?